test/powershell/engine/Basic/NativeCommandErrorHandling.Tests.ps1

```diff
@@ -0,0 +1,175 @@
+            $ErrorActionPreference = 'Continue'
+
+            $stderr = testexe -returncode 1 2>&1
+
+            $error[0].FullyQualifiedErrorId | Should -Be 'ProgramExitedWithNonZeroCode'
+            $stderr[1].Exception.Message | Should -Be "Program `"$exeName`" ended with non-zero exit code: 1."
+        }
+
+        It 'Non-zero exit code generates a non-teminating error for $ErrorActionPreference = ''SilentlyContinue''' {
+            $ErrorActionPreference = 'SilentlyContinue'
+
+            testexe -returncode 1 > $null
+
+            $error.Count | Should -Be 1
+            $error[0].FullyQualifiedErrorId | Should -Be 'ProgramExitedWithNonZeroCode'
+        }
+
+        It 'Non-zero exit code does not generates an error record for $ErrorActionPreference = ''Ignore''' {
+            $ErrorActionPreference = 'Ignore'
+
+            testexe -returncode 1 > $null
+
+            $LASTEXITCODE | Should -Be 1
+            $error.Count | Should -Be 0
+        }
+
+        It 'Zero exit code generates no error for $ErrorActionPreference = ''<ErrorActionPref>''' -TestCases $errorActionPrefTestCases {
+            param($ErrorActionPref)
+
+            $ErrorActionPreference = $ErrorActionPref
+
+            $output = testexe -returncode 0
+
+            $output | Should -BeExactly '0'
+            $LASTEXITCODE | Should -Be 0
+            $Error.Count | Should -Be 0
+        }
+
+        It 'Works as expected with a try/catch block when $ErrorActionPreference = ''<ErrorActionPref>''' -TestCase $errorActionPrefTestCases {
+            param($ErrorActionPref)
+
+            $ErrorActionPreference = $ErrorActionPref
+
+            $threw = $false
+            $continued = $false
+            $hitFinally = $false
+            try
+            {
+                testexe -returncode 17 2>&1 > $null
+                $continued = $true
+            }
+            catch
+            {
+                $threw = $true
+                $exception = $_.Exception
+            }
+            finally
+            {
+                $hitFinally = $true
+            }
+
+            $hitFinally | Should -BeTrue
+            $continued  | Should -Be ($ErrorActionPreference -ne 'Stop')
+            $threw      | Should -Be ($ErrorActionPreference -eq 'Stop')
+
+            if ($threw)
+            {
+                $exception.Path      | Should -BeExactly (Get-Command -Name testexe -CommandType Application).Path
+                $exception.ExitCode  | Should -Be $LASTEXITCODE
+                $exception.ProcessId | Should -BeGreaterThan 0
+            }
+        }
+
+        It 'Works with trap when $ErrorActionPreference = ''<ErrorActionPref>''' -TestCases $errorActionPrefTestCases {
+            param($ErrorActionPref)
+
+            $ErrorActionPreference = $ErrorActionPref
+
+            trap
+            {
+                $hitTrap = $true
+                $exception = $_
+                continue
+            }
+
+            $hitTrap = $false
```

Comment:
Should we explicitly create $exception in the scope too?

Reply:
Well it'll be `$null` by default, rather than `$false`, so we don't really need to